File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-1/version-1/javascript/worker.js
self.onmessage = async function (e) {
    const data = e.data;

    if (data.type === "do_analisis") {
        try {
            const { id, num_requests, delay } = data;
            const BASE_URL = location.origin;

            const urls = Array.from(
                { length: num_requests },
                () => `${BASE_URL}/api/4.1.1/${delay}.ts`
            );

            const results = [];
            const responseTimes = [];

            const fetchWithTiming = async (url) => {
                const start = performance.now();
                try {
                    const res = await fetch(url);
                    if (!res.ok) throw new Error(`Status ${res.status}`);

                    const payload = await res.json();
                    const elapsed = performance.now() - start;

                    results.push(payload);
                    responseTimes.push(elapsed);
                } catch (err) {
                    const elapsed = performance.now() - start;
                    results.push({ error: err.message });
                    responseTimes.push(elapsed);
                }
            };

            const tStart = performance.now();
            await Promise.all(urls.map(fetchWithTiming));
            const total_time = performance.now() - tStart;

            const totalResponseTime = responseTimes.reduce((sum, t) => sum + t, 0);
            const avg_time = num_requests > 0 ? totalResponseTime / num_requests : 0;

            const lastRes = results[results.length - 1];
            const lastArr = lastRes && Array.isArray(lastRes.data) ? lastRes.data : [];
            const last_value = lastArr.length > 0 ? lastArr[lastArr.length - 1].value : null;

            const result = {
                total_time,
                avg_time,
                last_value,
            };

            self.postMessage({
                id,
                json: JSON.stringify(result),
            });
        } catch (error) {
            self.postMessage({
                id: data.id,
                error: `Error in worker: ${error.message}`,
            });
        }
    }
};

File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-1/version-1/python/worker.py
from pyscript import fetch, sync
import asyncio
import time


async def do_requests(worker_time, num_requests, delay, base_url):
    urls = [f"{base_url}/api/4.1.1/{delay}.ts" for _ in range(num_requests)]
    start_time = time.perf_counter()
    individual_times = []

    async def fetch_url(url):
        t0 = time.perf_counter()
        resp = await fetch(url)
        elapsed = (time.perf_counter() - t0) * 1000
        individual_times.append(elapsed)
        if resp.status == 200:
            try:
                python_dict = await resp.json()
                return python_dict
            except Exception as e:
                return {"error": f"JSON parse failed: {str(e)}"}
        return {"error": f"Request failed with status {resp.status}"}

    results = await asyncio.gather(*(fetch_url(u) for u in urls))
    total_time = (time.perf_counter() - start_time) * 1000
    avg_time = sum(individual_times) / \
        len(individual_times) if individual_times else 0

    return {
        "worker_time": worker_time,
        "total_time": total_time,
        "avg_time": avg_time,
        "last_value": results[-1]["data"][-1]["value"],
    }


sync.do_requests = do_requests
